src/snowflake/connector/gcs_util.py

```diff
@@ -286,22 +308,61 @@ def get_file_header(meta, filename):
             url, we have no way of verifying that, except with the http status code of 200 which we have already
             confirmed and set the meta['result_status'] = UPLOADED/DOWNLOADED.
         """
-        if meta.get('presigned_url', None):
-            if meta.get('result_status', None) == ResultStatus.UPLOADED \
-                    or meta.get('result_status', None) == ResultStatus.DOWNLOADED:
-                return FileHeader(
-                    digest=meta.get(
-                        GCS_FILE_HEADER_DIGEST, None),
-                    content_length=meta.get(
-                        GCS_FILE_HEADER_CONTENT_LENGTH, None),
-                    encryption_metadata=meta.get(
-                        GCS_FILE_HEADER_ENCRYPTION_METADATA, None),
-                )
-            else:
-                meta['result_status'] = ResultStatus.NOT_FOUND_FILE
+        if meta.get('result_status', None) == ResultStatus.UPLOADED \
+                or meta.get('result_status', None) == ResultStatus.DOWNLOADED:
+            return FileHeader(
+                digest=meta.get(
+                    GCS_FILE_HEADER_DIGEST, None),
+                content_length=meta.get(
+                    GCS_FILE_HEADER_CONTENT_LENGTH, None),
+                encryption_metadata=meta.get(
+                    GCS_FILE_HEADER_ENCRYPTION_METADATA, None),
+            )
+        else:
+            meta['result_status'] = ResultStatus.NOT_FOUND_FILE
 
         return FileHeader(
             digest=None,
             content_length=None,
             encryption_metadata=None
         )
+
+    @staticmethod
+    def extract_bucket_name_and_path(stage_location):
+        stage_location = os.path.expanduser(stage_location)
+        container_name = stage_location
+        path = ''
+
+        # split stage location as bucket name and path
+        if '/' in stage_location:
+            container_name = stage_location[0:stage_location.index('/')]
+            path = stage_location[stage_location.index('/') + 1:]
+            if path and not path.endswith('/'):
+                path += '/'
+
+        return GcsLocation(
+            bucket_name=container_name,
+            path=path)
+
+    @staticmethod
+    def generate_file_url(stage_location, filename):
```

Comment:
Why do we generate the file URL? Are we not getting the file URLs from GS?

Reply:
GS only give us the access token in cased of downscope, so we have to build the url.